internal/observability/event/serialized_writer.go

```diff
@@ -0,0 +1,39 @@
+package event
+
+import (
+	"bytes"
+	"io"
+	"sync"
+
+	"github.com/hashicorp/boundary/internal/errors"
+)
+
+// serializedWriter uses a mutext to serializes all writes to its io.Writer
+type serializedWriter struct {
+	l *sync.Mutex
+	w io.Writer
+}
+
+// Write uses a mutex to serialize all writes
+func (s *serializedWriter) Write(p []byte) (int, error) {
+	const op = "event.(serializedWriter).Write"
+	if s == nil {
+		return 0, errors.New(errors.InvalidParameter, op, "missing serialized writer")
+	}
+	if s.l == nil {
+		return 0, errors.New(errors.InvalidParameter, op, "missing lock")
+	}
+	if s.w == nil {
+		return 0, errors.New(errors.InvalidParameter, op, "missing writer")
+	}
+
+	s.l.Lock()
+	defer s.l.Unlock()
+	reader := bytes.NewReader(p)
+
+	n, err := reader.WriteTo(s.w)
+	if err != nil {
+		return 0, err
+	}
+	return int(n), err
```

Comment:
Rather than cast this, why not return int64? All `io.Reader`-ish functions use that so it seems more standard.

Reply:
the interface requires an int
```
type Writer interface {
    Write(p []byte) (n int, err error)
}
```

Reply:
Wow. TIL that Writer uses int and Reader uses int64. I never really noticed it before. I guess they figure you're never going to manage to write more than 2^32 (except when it's 2^16 or 2^64) bytes in one call. Still...bizarre.